{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collection date aggregation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import copy\n",
    "import numpy as np\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data loading"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Right after loading the metadata, we separate it into genomics and non-genomics, because we are considring genomics metadata as time-independent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_parquet(\"processed_metadata.parquet\")\n",
    "genomics_rows = metadata[metadata[\"omics\"].str.contains(\"genomics\") == True]\n",
    "non_genomics_rows = metadata[metadata[\"omics\"].str.contains(\"genomics\") == False]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample collection data collapse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To prepare the data for collapse:\n",
    "1. group by unique patient\n",
    "\n",
    "*For each patient:*\n",
    "\n",
    "2. sort by `sample_collected_date`\n",
    "3. calculate the difference (in days) between the current sample and the previous one\n",
    "4. assign 0 to the first sample on each patient\n",
    "\n",
    "*Reaggregate the patients:*\n",
    "\n",
    "5. concatenate the ordered patient tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_groupings = non_genomics_rows.groupby(\"patient_id\")\n",
    "ordered_list = []\n",
    "for id, current_df in patient_groupings:\n",
    "    sorted_by_date_df = current_df.sort_values(\"sample_collected_date\", ascending=True)\n",
    "    sorted_by_date_df[\"days_since_last_sample\"] = (\n",
    "        sorted_by_date_df[\"sample_collected_date\"].diff().dt.days\n",
    "    )\n",
    "\n",
    "    sorted_by_date_df[\"days_since_last_sample\"] = (\n",
    "        sorted_by_date_df[\"days_since_last_sample\"].fillna(0).astype(int)\n",
    "    )\n",
    "    ordered_list.append(sorted_by_date_df)\n",
    "\n",
    "ordered_df = pd.concat(ordered_list, axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Investigate single patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>omics</th>\n",
       "      <th>sample_collected_date</th>\n",
       "      <th>days_since_last_sample</th>\n",
       "      <th>simple_tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>9032329</td>\n",
       "      <td>FS01338644</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2017-10-19</td>\n",
       "      <td>0</td>\n",
       "      <td>colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1102</th>\n",
       "      <td>9032329</td>\n",
       "      <td>FS01629014</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2017-10-19</td>\n",
       "      <td>0</td>\n",
       "      <td>small_intestine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1111</th>\n",
       "      <td>9032329</td>\n",
       "      <td>10521576</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>2017-10-19</td>\n",
       "      <td>0</td>\n",
       "      <td>plasma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1094</th>\n",
       "      <td>9032329</td>\n",
       "      <td>10641613</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>2018-09-06</td>\n",
       "      <td>322</td>\n",
       "      <td>plasma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104</th>\n",
       "      <td>9032329</td>\n",
       "      <td>FS04156255</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2018-09-06</td>\n",
       "      <td>0</td>\n",
       "      <td>small_intestine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1113</th>\n",
       "      <td>9032329</td>\n",
       "      <td>FS04156258</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2018-09-06</td>\n",
       "      <td>0</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1120</th>\n",
       "      <td>9032329</td>\n",
       "      <td>FS01694799</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2018-09-06</td>\n",
       "      <td>0</td>\n",
       "      <td>colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100</th>\n",
       "      <td>9032329</td>\n",
       "      <td>FR20600048</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>2019-05-30</td>\n",
       "      <td>266</td>\n",
       "      <td>plasma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1110</th>\n",
       "      <td>9032329</td>\n",
       "      <td>8051600906</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2019-05-30</td>\n",
       "      <td>0</td>\n",
       "      <td>colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1121</th>\n",
       "      <td>9032329</td>\n",
       "      <td>8051600883</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2019-05-30</td>\n",
       "      <td>0</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1098</th>\n",
       "      <td>9032329</td>\n",
       "      <td>8051600870</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2019-06-27</td>\n",
       "      <td>28</td>\n",
       "      <td>small_intestine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1117</th>\n",
       "      <td>9032329</td>\n",
       "      <td>8051600894</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2019-06-27</td>\n",
       "      <td>0</td>\n",
       "      <td>colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1097</th>\n",
       "      <td>9032329</td>\n",
       "      <td>8051384122</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2020-07-02</td>\n",
       "      <td>371</td>\n",
       "      <td>colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1099</th>\n",
       "      <td>9032329</td>\n",
       "      <td>FR20602476</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>2020-07-02</td>\n",
       "      <td>0</td>\n",
       "      <td>plasma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1115</th>\n",
       "      <td>9032329</td>\n",
       "      <td>8051384121</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>2020-07-02</td>\n",
       "      <td>0</td>\n",
       "      <td>small_intestine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      patient_id   sample_id            omics sample_collected_date  \\\n",
       "1095     9032329  FS01338644  transcriptomics            2017-10-19   \n",
       "1102     9032329  FS01629014  transcriptomics            2017-10-19   \n",
       "1111     9032329    10521576       proteomics            2017-10-19   \n",
       "1094     9032329    10641613       proteomics            2018-09-06   \n",
       "1104     9032329  FS04156255  transcriptomics            2018-09-06   \n",
       "1113     9032329  FS04156258  transcriptomics            2018-09-06   \n",
       "1120     9032329  FS01694799  transcriptomics            2018-09-06   \n",
       "1100     9032329  FR20600048       proteomics            2019-05-30   \n",
       "1110     9032329  8051600906  transcriptomics            2019-05-30   \n",
       "1121     9032329  8051600883  transcriptomics            2019-05-30   \n",
       "1098     9032329  8051600870  transcriptomics            2019-06-27   \n",
       "1117     9032329  8051600894  transcriptomics            2019-06-27   \n",
       "1097     9032329  8051384122  transcriptomics            2020-07-02   \n",
       "1099     9032329  FR20602476       proteomics            2020-07-02   \n",
       "1115     9032329  8051384121  transcriptomics            2020-07-02   \n",
       "\n",
       "      days_since_last_sample    simple_tissue  \n",
       "1095                       0            colon  \n",
       "1102                       0  small_intestine  \n",
       "1111                       0           plasma  \n",
       "1094                     322           plasma  \n",
       "1104                       0  small_intestine  \n",
       "1113                       0            other  \n",
       "1120                       0            colon  \n",
       "1100                     266           plasma  \n",
       "1110                       0            colon  \n",
       "1121                       0            other  \n",
       "1098                      28  small_intestine  \n",
       "1117                       0            colon  \n",
       "1097                     371            colon  \n",
       "1099                       0           plasma  \n",
       "1115                       0  small_intestine  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_example_no_genomics = ordered_df[ordered_df[\"patient_id\"] == 9032329]\n",
    "subset_example_no_genomics[\n",
    "    [\n",
    "        \"patient_id\",\n",
    "        \"sample_id\",\n",
    "        \"omics\",\n",
    "        \"sample_collected_date\",\n",
    "        \"days_since_last_sample\",\n",
    "        \"simple_tissue\",\n",
    "    ]\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aggregate omic types by interval"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next steps are deployed jointly and encompass:\n",
    "- Identifying unique 1-week intervals, per patient, with respective multi-omics information\n",
    "- Aggregating the omics types in these periods\n",
    "- Assigning the tissues to the interval information, according to transcriptomics, as transcriptomics data can have multiple different tissues\n",
    "- Reconcatenating genomics information according to the `patient_id`, since we are considering genomics data as independent from collection date\n",
    "- Each sample is now described by the `patient_id` and the `interval_id` and is characterized by the `transcriptomics`, `proteomics`, `genomics_exome` and `genomics_array` columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/1970 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1970/1970 [01:51<00:00, 17.72it/s]\n"
     ]
    }
   ],
   "source": [
    "def process_interval_rows(input_rows, interval_id):\n",
    "    interval_df = pd.DataFrame(input_rows)\n",
    "    interval_df[\"interval_id\"] = (\n",
    "        str(interval_id) + \"_\" + interval_df[\"simple_tissue\"].astype(str)\n",
    "    )\n",
    "\n",
    "    # Proteomics data is plasma, so we set it aside for now\n",
    "    tissue_proteomics_df = interval_df[interval_df[\"omics\"] == \"proteomics\"]\n",
    "    unique_interval_ids = interval_df[\"interval_id\"].unique()\n",
    "    tissue_intervals = []\n",
    "    for current_tissue_interval in unique_interval_ids:\n",
    "        tissue_subset = interval_df[\n",
    "            interval_df[\"interval_id\"] == current_tissue_interval\n",
    "        ]\n",
    "\n",
    "        tissue_transcriptomics_df = tissue_subset[\n",
    "            tissue_subset[\"omics\"] == \"transcriptomics\"\n",
    "        ]\n",
    "\n",
    "        # Pick only the first proteomics and transcriptomics sample\n",
    "        if tissue_proteomics_df.shape[0] >= 1:\n",
    "            proteomics_row = copy.deepcopy(tissue_proteomics_df.iloc[0])\n",
    "\n",
    "            # We reassign the tissue with the corresponding transcriptomics, to later match the ids\n",
    "            proteomics_row[\"interval_id\"] = current_tissue_interval\n",
    "            tissue_intervals.append(proteomics_row)\n",
    "\n",
    "        elif tissue_proteomics_df.shape[0] == 0:\n",
    "            pass\n",
    "        if tissue_transcriptomics_df.shape[0] >= 1:\n",
    "            tissue_intervals.append(tissue_transcriptomics_df.iloc[0])\n",
    "        elif tissue_transcriptomics_df.shape[0] == 0:\n",
    "            pass\n",
    "    return pd.DataFrame(tissue_intervals)\n",
    "\n",
    "\n",
    "def process_single_patient(input_df):\n",
    "    interval_grouping, interval_rows, interval_id = [], [], 1\n",
    "    for index, row in input_df.iterrows():\n",
    "        if row[\"days_since_last_sample\"] > 7:\n",
    "            cleaner_interval_df = process_interval_rows(interval_rows, interval_id)\n",
    "            interval_grouping.append(cleaner_interval_df)\n",
    "            interval_rows = []\n",
    "            interval_id += 1\n",
    "        interval_rows.append(row)\n",
    "\n",
    "    # Don't forget the last interval\n",
    "    cleaner_interval_df = process_interval_rows(interval_rows, interval_id)\n",
    "    interval_grouping.append(cleaner_interval_df)\n",
    "\n",
    "    return pd.concat(interval_grouping, axis=0)\n",
    "\n",
    "\n",
    "def flatten_intervals(input_df, genomics_rows):\n",
    "    # If there is no genomics data, assign NaN identifiers\n",
    "    try:\n",
    "        genomics_exome_id = genomics_rows[genomics_rows[\"omics\"] == \"genomics_exome\"][\n",
    "            \"sample_id\"\n",
    "        ].values[0]\n",
    "        genomics_array_id = genomics_rows[genomics_rows[\"omics\"] == \"genomics_array\"][\n",
    "            \"sample_id\"\n",
    "        ].values[0]\n",
    "    except:\n",
    "        genomics_exome_id = np.nan\n",
    "        genomics_array_id = np.nan\n",
    "\n",
    "    unique_intervals = input_df[\"interval_id\"].unique()\n",
    "    all_intervals = []\n",
    "    for current_interval in unique_intervals:\n",
    "        current_subset = copy.deepcopy(\n",
    "            input_df[input_df[\"interval_id\"] == current_interval]\n",
    "        )\n",
    "\n",
    "        # If there are two samples in the interval, they must be from transcriptomics and proteomics\n",
    "        if current_subset.shape[0] == 2:\n",
    "            current_subset.sort_values(\n",
    "                by=\"omics\", ascending=False, inplace=True\n",
    "            )  # Make sure transcriptomics appears first\n",
    "            comparison_result = copy.deepcopy(\n",
    "                current_subset.iloc[0].combine_first(current_subset.iloc[1])\n",
    "            )\n",
    "            comparison_result[\"transcriptomics\"] = current_subset.iloc[0][\"sample_id\"]\n",
    "            comparison_result[\"proteomics\"] = current_subset.iloc[1][\"sample_id\"]\n",
    "\n",
    "        # If there is only one sample in the interval, it must be either transcriptomics or proteomics\n",
    "        elif current_subset.shape[0] == 1:\n",
    "            comparison_result = copy.deepcopy(current_subset.iloc[0])\n",
    "\n",
    "            # If the sample is transcriptomics, assign NaN to proteomics\n",
    "            if comparison_result[\"omics\"] == \"transcriptomics\":\n",
    "                comparison_result[\"transcriptomics\"] = comparison_result[\"sample_id\"]\n",
    "                comparison_result[\"proteomics\"] = np.nan\n",
    "\n",
    "            # If the sample is proteomics, assign NaN to transcriptomics\n",
    "            elif comparison_result[\"omics\"] == \"proteomics\":\n",
    "                comparison_result[\"transcriptomics\"] = np.nan\n",
    "                comparison_result[\"proteomics\"] = comparison_result[\"sample_id\"]\n",
    "        comparison_result[\"genomics_exome\"] = genomics_exome_id\n",
    "        comparison_result[\"genomics_array\"] = genomics_array_id\n",
    "        all_intervals.append(comparison_result)\n",
    "    return pd.DataFrame(all_intervals)\n",
    "\n",
    "\n",
    "complete_list = []\n",
    "for current_patient in tqdm(list(ordered_df[\"patient_id\"].unique())):\n",
    "    interval_grouping_df = process_single_patient(\n",
    "        ordered_df[ordered_df[\"patient_id\"] == current_patient]\n",
    "    )\n",
    "    complete_list.append(\n",
    "        flatten_intervals(\n",
    "            interval_grouping_df,\n",
    "            genomics_rows[genomics_rows[\"patient_id\"] == current_patient],\n",
    "        )\n",
    "    )\n",
    "\n",
    "complete_df = pd.concat(complete_list, axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `sample_id` and the `omics` columns no longer make sense anymore as the `interval_id` is now the unique identifier and the omics columns has been pivoted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>interval_id</th>\n",
       "      <th>transcriptomics</th>\n",
       "      <th>proteomics</th>\n",
       "      <th>genomics_exome</th>\n",
       "      <th>genomics_array</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>9032329</td>\n",
       "      <td>1_colon</td>\n",
       "      <td>FS01338644</td>\n",
       "      <td>10521576</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1102</th>\n",
       "      <td>9032329</td>\n",
       "      <td>1_small_intestine</td>\n",
       "      <td>FS01629014</td>\n",
       "      <td>10521576</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1111</th>\n",
       "      <td>9032329</td>\n",
       "      <td>1_plasma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10521576</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1094</th>\n",
       "      <td>9032329</td>\n",
       "      <td>2_plasma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10641613</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104</th>\n",
       "      <td>9032329</td>\n",
       "      <td>2_small_intestine</td>\n",
       "      <td>FS04156255</td>\n",
       "      <td>10641613</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1113</th>\n",
       "      <td>9032329</td>\n",
       "      <td>2_other</td>\n",
       "      <td>FS04156258</td>\n",
       "      <td>10641613</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1120</th>\n",
       "      <td>9032329</td>\n",
       "      <td>2_colon</td>\n",
       "      <td>FS01694799</td>\n",
       "      <td>10641613</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1100</th>\n",
       "      <td>9032329</td>\n",
       "      <td>3_plasma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FR20600048</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1110</th>\n",
       "      <td>9032329</td>\n",
       "      <td>3_colon</td>\n",
       "      <td>8051600906</td>\n",
       "      <td>FR20600048</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1121</th>\n",
       "      <td>9032329</td>\n",
       "      <td>3_other</td>\n",
       "      <td>8051600883</td>\n",
       "      <td>FR20600048</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1098</th>\n",
       "      <td>9032329</td>\n",
       "      <td>4_small_intestine</td>\n",
       "      <td>8051600870</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1117</th>\n",
       "      <td>9032329</td>\n",
       "      <td>4_colon</td>\n",
       "      <td>8051600894</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1097</th>\n",
       "      <td>9032329</td>\n",
       "      <td>5_colon</td>\n",
       "      <td>8051384122</td>\n",
       "      <td>FR20602476</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1099</th>\n",
       "      <td>9032329</td>\n",
       "      <td>5_plasma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FR20602476</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1115</th>\n",
       "      <td>9032329</td>\n",
       "      <td>5_small_intestine</td>\n",
       "      <td>8051384121</td>\n",
       "      <td>FR20602476</td>\n",
       "      <td>1203065541</td>\n",
       "      <td>203990550113_R06C02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      patient_id        interval_id transcriptomics  proteomics  \\\n",
       "1095     9032329            1_colon      FS01338644    10521576   \n",
       "1102     9032329  1_small_intestine      FS01629014    10521576   \n",
       "1111     9032329           1_plasma             NaN    10521576   \n",
       "1094     9032329           2_plasma             NaN    10641613   \n",
       "1104     9032329  2_small_intestine      FS04156255    10641613   \n",
       "1113     9032329            2_other      FS04156258    10641613   \n",
       "1120     9032329            2_colon      FS01694799    10641613   \n",
       "1100     9032329           3_plasma             NaN  FR20600048   \n",
       "1110     9032329            3_colon      8051600906  FR20600048   \n",
       "1121     9032329            3_other      8051600883  FR20600048   \n",
       "1098     9032329  4_small_intestine      8051600870         NaN   \n",
       "1117     9032329            4_colon      8051600894         NaN   \n",
       "1097     9032329            5_colon      8051384122  FR20602476   \n",
       "1099     9032329           5_plasma             NaN  FR20602476   \n",
       "1115     9032329  5_small_intestine      8051384121  FR20602476   \n",
       "\n",
       "     genomics_exome       genomics_array  \n",
       "1095     1203065541  203990550113_R06C02  \n",
       "1102     1203065541  203990550113_R06C02  \n",
       "1111     1203065541  203990550113_R06C02  \n",
       "1094     1203065541  203990550113_R06C02  \n",
       "1104     1203065541  203990550113_R06C02  \n",
       "1113     1203065541  203990550113_R06C02  \n",
       "1120     1203065541  203990550113_R06C02  \n",
       "1100     1203065541  203990550113_R06C02  \n",
       "1110     1203065541  203990550113_R06C02  \n",
       "1121     1203065541  203990550113_R06C02  \n",
       "1098     1203065541  203990550113_R06C02  \n",
       "1117     1203065541  203990550113_R06C02  \n",
       "1097     1203065541  203990550113_R06C02  \n",
       "1099     1203065541  203990550113_R06C02  \n",
       "1115     1203065541  203990550113_R06C02  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaner_table = complete_df.drop([\"sample_id\", \"omics\"], axis=1)\n",
    "cleaner_table[cleaner_table[\"patient_id\"] == 9032329][\n",
    "    [\n",
    "        \"patient_id\",\n",
    "        \"interval_id\",\n",
    "        \"transcriptomics\",\n",
    "        \"proteomics\",\n",
    "        \"genomics_exome\",\n",
    "        \"genomics_array\",\n",
    "    ]\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaner_table.to_parquet(\"collapsed_metadata.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original metadata shape: (11270, 143) and final shape: (5309, 149)\n"
     ]
    }
   ],
   "source": [
    "print(f\"Original metadata shape: {metadata.shape} and final shape: {complete_df.shape}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
